distributed/worker_state_machine.py

```diff
@@ -3650,8 +3650,11 @@ def _finish_async_instruction(
             logger.exception("async instruction handlers should never raise!")
             raise
 
-        with ledger.record():
-            # Capture metric events in _transition_to_memory()
+        # Capture metric events in _transition_to_memory()
+        # As this may trigger calls to _start_async_instruction for more tasks,
+        # make sure we don't endlessly pile up context_meter callbacks by specifying
+        # the same key as in _start_async_instruction.
+        with ledger.record(key="async-instruction"):
```

Comment:
As an aside, this whole incident would have been prevented by 
- https://github.com/dask/distributed/pull/7686